lang/syn/src/codegen/program/handlers.rs

```diff
@@ -91,6 +91,16 @@ pub fn generate(program: &Program) -> proc_macro2::TokenStream {
         }
     };
 
+    let non_inlined_event: proc_macro2::TokenStream = {
+        quote! {
+            #[inline(never)]
+            #[cfg(not(feature = "no-cpi-events"))]
+            pub fn __event_dispatch(program_id: &Pubkey, accounts: &[AccountInfo], event_data: &[u8]) -> anchor_lang::Result<()> {
```

Comment:
We are not validating anything here, anyone can call this instruction.

Reply:
Added validation in https://github.com/coral-xyz/anchor/pull/2438/commits/87652d2f90ae00788482fbd540aff3d8be56a386.